src/app/src/main/kotlin/org/imperial/mrc/hint/controllers/ADRController.kt

```diff
@@ -201,20 +204,50 @@ class ADRController(private val encryption: Encryption,
 
         // 4. Checksum file and upload with metadata to ADR
         val filePart = Pair("upload", file)
+        val fileHash = file.md5sum()
         val commonParameters =
-                listOf("name" to resourceFileName, "description" to artefact.second, "hash" to file.md5sum(),
+                listOf("name" to resourceFileName, "description" to artefact.second, "hash" to fileHash,
                         "resource_type" to resourceType)
         val adr = adrClientBuilder.build()
         return try
         {
             when (resourceId)
             {
                 null -> adr.postFile("resource_create", commonParameters + listOf("package_id" to id), filePart)
-                else -> adr.postFile("resource_patch", commonParameters + listOf("id" to resourceId), filePart)
+                else ->
+                {
+                    val fileHasChangesResponse = uploadFileHasChanges(resourceId, fileHash)
+                    if (fileHasChangesResponse.statusCode.is2xxSuccessful && fileHasChangesResponse.body!!)
+                    {
+                        adr.postFile("resource_patch", commonParameters + listOf("id" to resourceId), filePart)
+                    }
+                    else
+                    {
+                        EmptySuccessResponse.asResponseEntity()
+                    }
+                }
             }
         }
         finally {
             tmpDir.deleteRecursively()
         }
     }
+
+    fun uploadFileHasChanges(resourceId: String, newDatasetHash: String): ResponseEntity<Boolean>
```

Comment:
Hi @carbowizzy. I think this an improvement but it has also gotten a bit difficult to follow: particularly constructing a response just to indicate success or failure to another method. I suggest something like this:
```kotlin
    fun uploadFileHasChanges(resourceId: String, newDatasetHash: String): Boolean
    {
        val adr = adrClientBuilder.build()
        val response = adr.get("resource_show?id=${resourceId}")
        if (response.statusCode.isError) {
            throw IOException("Unable to retrieve hash")
        }
        val parser = JSONParser()
        val json: JSONObject = parser.parse(response.body!!) as JSONObject
        val data: JSONObject = json["data"] as JSONObject
        return data["hash"] != newDatasetHash
    }
```

Then you can handle in the method above as:
```kotlin
        return try
        {
            when (resourceId)
            {
                null -> adr.postFile("resource_create", commonParameters + listOf("package_id" to id), filePart)
                else ->
                {
                    if (uploadFileHasChanges(resourceId, fileHash))
                    {
                        adr.postFile("resource_patch", commonParameters + listOf("id" to resourceId), filePart)
                    }
                    else
                    {
                        EmptySuccessResponse.asResponseEntity()
                    }
                }
            }
        }
        catch (e: IOException) {
            ErrorDetail(HttpStatus.INTERNAL_SERVER_ERROR, e.message!!).toResponseEntity()
        }
        finally
        {
            tmpDir.deleteRecursively()
        }
```